Shapemap Tools
==============
A set of tools to manage shapemap on RNA families.

## Installation 

Prior to install shapemap-tools you must install miniconda or micromamba.

Using git repository :

```bash
git clone https://github.com/Sargueil-CiTCoM/shapemap-tools.git 
cd shapemap-tools
conda create --name shapemap-tools --file conda-env.yml
pip install -e .
```

Via conda (not yet available)

```bash
conda env create shapemap-tools
conda activate shapemap-tools
conda install shapemap-tools -c conda-forge -c bioconda
```

## Usage 

All script are prefixed by `shpm_*`

To start using scripts, you must activate your environnement
```bash
conda activate shapemap-tools
```

An example data treatment script is provided in the example folder.


## Library preparation
## GenTag 

```bash
shpm_gentag --sequences [sequence to tag in fasta] --primers [sequences of primers in fasta] --tags [tags to use in fasta] --prefix [prefix for sequence ids] --output=[Output file].tsv

```

Example command :
```bash
shpm_gentag --sequences data/2021-09-13-riboswitch-sam.fasta --primers data/2021-08-31-primer-aptamer-sam.fasta --tags data/barcodes-10bp-4dist-ashlock.fasta --prefix RIBOSAMN --output=2021-09-15-tagged-riboswitch-sam.tsv
```

Example 5S :
```bash
shpm_gentag --sequences data/2021-09-15_5SrRNA_RF00001-DNA.fasta --primers data/2021-10-11-primer-5SrRNA.fasta --tags data/barcodes-10bp-5dist-ashlock.fasta --prefix rRNA5S --output 2021-10-11-tagged-5SrRNA.tsv
```

```bash
shpm_gentag --sequences data/2021-12-10_SAMAP-PDB-Seq.fasta --primers data/2021-08-31-primer-aptamer-sam.fasta --tags data/barcodes-10bp-5dist-ashlock.fasta --prefix SAMAP-PDB --output output/2021-12-10-tagged-SAMAP-PDB.tsv --remainingtags output/2021-12-10_remain-after-1st-design

```


## Treatment
### Demultiplexing - cutadapt

Demultiplex fastq reads using tagging system generated by shpm_gentag

```bash
shpm_demultiplex [TSV file from gentag] --cores [nthreads to use] --max-errors 1 --output_path [Output folder for demultiplexed reads]  --folder [fastq files folder]
```

More information with :

```bash
shpm_demultiplex --help
```

### ShapeMapper Launcher

Run shapemapper on multiple conditions using configurations files 

Note : Shapemapper 2.1.5 must be installed from a release (compiled with third party dependency). 

Official release : https://github.com/Weeks-UNC/shapemapper2/releases/tag/2.1.5

Some bug correction have been made in this repository : https://github.com/Sargueil-CiTCoM/shapemapper2

you can patch the official release with those modifications.


```bash
shpm_launch_shapemapper config/decrypted-SAMAP_config.yaml config/decrypted-SAMAP_samples.tsv  --shapemapper-path [path to shapemapper dir]/shapemapper --dnerase=True --verbose=True --nthreads=48

```

samples.tsv: Contains the list of conditions
config.yaml: contains information about shapemapper process

In order for this script to work, you must provide using `--shapemapper-path` the path to shapemapper binary. So you must install shapemapper on your computer.

Examples config files :
![config.yaml](example/decrypted/config/decrypted-SAMAP_config.yaml)
![samples.tsv](example/decrypted/config/decrypted-SAMAP_samples.tsv)


More information with :
```bash
shpm_launch_shapemapper --help
```

## Normalize By conditions

Wrapper around Shapemapper normalization script in order to easy normalize a RNA for all condition

Example:
```bash
shpm_norm_by_conditions [input_folder] [output-folder] --shapemapper-path [path to shapemapper dir] --condition-prefix [prefix of molecules] --mode [all, conditions, conditions-reps] --render-figure=False --nthreads=48
```


More information with :
```bash
shpm_norm_by_conditions --help
```


## Aggregate replicates

Aggregate multiple replicate of a given condition for all RNA of a family, with divers statistics
then, plot

```bash
shpm_agg_replicates --input_path [input_reactivity_folder] --output_path [output_aggregation_folder] --config_path config/config.yaml
shpm_plot_aggregate [output_aggregation_folder] --nthreads=48
```


## Generate structure

Wrapper around IPANEMAP in order to generate structure for each RNA of a family

```bash
shpm_gen_structures [input_reactivity_folder] [outputfolder] [input_sequences] --nthreads=24
```

## Foorprint

Compare and interpret chemical reactivity maps obtained in different conditions

```bash
shpm_footprint [folder_containing_conditions] [condition1] [condition2] --nthreads=24
```

## Data Visualization

### Result viewer

A tool to observe and compare member of a RNA family condition by conditions.

```bash
shpm_results_viewer  [folder_containing_conditions] [prefix_of_files_for_this_family]
```


### Quality report 

Assess quality of the data treated with `shpm_launch_shapemapper`

```bash
shpm_quality_report [folder_containing_conditions]
```

### Quality report compare

Compare the quality of two conditions regarding : 
- Number of read of each RNA
- Reactivity of each nucleotide of the same RNA molecule.

```bash
shpm_quality_report [folder_of_the_first_condition] [folder_of_second_condition]
```
